witness/src/Dscp/Witness/Retranslate/Transaction.hs

```diff
@@ -0,0 +1,59 @@
+{-# LANGUAGE OverloadedLists #-}
+
+module Dscp.Witness.Retranslate.Transaction
+    (
+    ) where
+
+import Control.Concurrent (threadDelay)
+import Control.Concurrent.STM (TQueue, TVar, modifyTVar, newTQueueIO, readTQueue, readTVar,
+                               writeTQueue, writeTVar)
+
+import qualified Data.HashMap.Strict as HashMap
+
+import Fmt ((+|), (+||), (|+), (||+))
+
+import Loot.Base.HasLens (HasLens (..))
+import Loot.Log (logError, logInfo)
+import Loot.Network.Class (ClientEnv, getPeers)
+import Loot.Network.ZMQ (ZmqTcp)
+
+import Dscp.Core
+import Dscp.Crypto
+import Dscp.Network.Wrapped
+import Dscp.Witness.Block.Logic
+import Dscp.Witness.Launcher.Mode
+import Dscp.Witness.Mempool
+import Dscp.Witness.Messages
+
+makeRetranslator :: forall ctx m. WitnessWorkMode ctx m => IO (Listener m, Worker m)
+makeRetranslator = do
+    localState <- newTVarIO HashMap.empty
+    queue      <- newTQueueIO
+
+    return (mkListener localState queue, mkWorker localState queue)
+  where
+    mkListener failedTxs queue =
+        Listener "transactionRetranslatorOutput" [] $ \btq -> do
+            dieGracefully $ do
+                forever $ do
+                    tx <- atomically $ readTQueue queue
+                    atomically $ servPub btq (RetranslateTx tx)
+
+    mkWorker failedTxs queue =
+        Worker "transactionRetranslatorInput" [] [subType @RetranslateTx] $ \btq -> do
+            dieGracefully $ do
+                forever $ do
+                    (_, RetranslateTx tx) <- cliRecvUpdate btq (-1)
+                    hashmap               <- atomically $ readTVar failedTxs
+
+                    unless (hash tx `elem` hashmap) $ do
+                        pool    <- view (lensOf @MempoolVar @ctx)
+                        isThere <- isInMempool pool tx
+
+                        unless isThere $ do
+                            addTxToMempool pool tx
+                            atomically $ writeTQueue queue tx
+
+    dieGracefully action =
+        action `catchAny` \e -> do
+            logError $ fromString $ "Exception in transactionRetranslatorInput: " <> show e
```

Comment:
I think this would better be handled in code which runs workers and listeners - it has to do that anyway, and do in uniform way.

Reply:
I copied it from some `Worker`. I thought it is there to catch immidately and not loose it somewhere between threads. Does our logger catch threads dying, by the way?

Reply:
I'm sure it doesn't, that would be a magic of some sort )